lib/ModelMixins/TileErrorHandlerMixin.ts

```diff
@@ -0,0 +1,366 @@
+      minTimeout: 200,
+      randomize: true
+    };
+
+    /**
+     * A hook that may be implemented by catalog items for custom handling of tile errors.
+     *
+     * @param maybeError A tile request promise that that fails with the tile error
+     * @param tile       The tile to be fetched
+     * @returns          A modified promise
+     *
+     * The item can then decide to retry the tile request after adding custom parameters
+     * like an authentication token.
+     */
+    handleTileError?: (
+      request: Promise<void>,
+      tile: { x: number; y: number; level: number }
+    ) => Promise<void>;
+
+    get hasTileErrorHandlerMixin() {
+      return true;
+    }
+
+    /*
+     * Handling tile errors is really complicated because:
+     *
+     * 1) things go wrong for a variety of weird reasons including server
+     * misconfiguration, servers that are flakey but not totally broken,
+     * etc.
+     *
+     * 2) we want to fail as gracefully as possible, and give flakey servers
+     * every chance chance to shine
+     *
+     * 3) we don't generally have enough information the first time something fails.
+     *
+     * There are several mechanisms in play here:
+     *   - Cesium's Resource automatically keeps trying to load any resource
+     *     that fails until told to stop, but tells us each time.
+     *   - The "retry" library knows how to pace the retries, and when to actually stop.
+     */
+    onTileLoadError(tileProviderError: TileProviderError): void {
+      const operation = retry.operation(this.tileRetryOptions);
+      // result.reject = stop trying to load this tile
+      // result.resolve = retry loading this tile
+      const result = when.defer();
+      const imageryProvider = tileProviderError.provider as ImageryProvider;
+      const tile = {
+        x: tileProviderError.x,
+        y: tileProviderError.y,
+        level: tileProviderError.level
+      };
+
+      // We're only concerned about failures for tiles that actually overlap this item's extent.
+      if (
+        isTileOutsideExtent(
+          tile,
+          runInAction(() => this.rectangle),
+          imageryProvider
+        )
+      ) {
+        tileProviderError.retry = false;
+        return;
+      }
+
+      // By setting retry to a promise, we tell cesium/leaflet to
+      // reload the tile if the promise resolves successfully
+      // https://github.com/TerriaJS/cesium/blob/terriajs/Source/Core/TileProviderError.js#L161
+      tileProviderError.retry = result;
+
+      if (tileProviderError.timesRetried === 0) {
+        // There was an intervening success, so restart our count of the tile failures.
+        this.tileFailures = 0;
+        this.tileRetriesByMap.clear();
+      }
+
+      operation.attempt(
+        action(async attemptNumber => {
+          if (this.show === false) {
+            // If the layer is no longer shown, ignore errors and don't retry.
+            tellMapToSilentlyGiveUp();
```

Comment:
I get an uncaught exception saying 
````
TypeError: tellMapToSilentlyGiveUp is not a function
````

Reply:
I have relocated the helper method definitions before its usage.